File: cpp/216.cpp
#include <vector>
using namespace std;

class Solution {
  void dfs(int k, int n, int i, vector<int> &cur, vector<vector<int>> &res){
    if(k == 0 && n == 0){
      res.push_back(cur);
      return;
    }
    if(k < 0 || n < 0 || i > 9)
      return;
    cur.push_back(i);
    dfs(k-1, n-i, i+1, cur, res);
    cur.pop_back();
    dfs(k, n, i+1, cur, res);
  }
public:
  vector<vector<int>> combinationSum3(int k, int n) {
    vector<vector<int>> res;
    vector<int> cur;
    dfs(k, n, 1, cur, res);
    return res;
  }
};

File: python/216.py
class Solution(object):

    def dfs(self, k, n, i, cur, res):
        if n == 0 and k == 0:
            res.append(cur[:])
            return
        if k < 0 or n < 0 or i > 9:
            return
        cur.append(i)
        self.dfs(k-1, n-i, i+1, cur, res)
        cur.pop()
        self.dfs(k, n, i+1, cur, res)

    def combinationSum3(self, k, n):
        """
        :type k: int
        :type n: int
        :rtype: List[List[int]]
        """
        res = []
        self.dfs(k, n, 1, [], res)
        return res

print(Solution().combinationSum3(3, 15))
